### PR TITLE
eng-557 build centralized linting system for all repos

### LINKED ISSUE #7: Create centralized linting system that can be applied to all MatrixAI repos, regardless of project-specific linting rules
### Specification

If we can create a single config bundle that includes all of the rules from MatrixAI codebases, we can greatly streamline the process of setting up linting rules in new repos.

### Tasks

1. Research into how to implement a bundled config file that can be used to apply rules from other eslint-plugins 
2. Create a bundled config file that combines all of the commonly used rules in our repos in a single package

### PR DESCRIPTION
**Pull request by CDeltakai**
### Description

This PR will implement a centralized linting system which will allow all MatrixAI repos to import from a single linting package (this one) and apply appropriate linting rules without further intervention. It should also allow downstream repos to extend the system in order to ignore certain rules or add project-specific rules.

### Issues Fixed

* Relates #7 

#### 🔧 Repository Refactor
- [x] 1. Rename the repository from `js-eslint` to `js-lint`
- [x] 2. Update `package.json`:
  - [x] 2.1 Change package name to `@matrixai/lint`
  - [x] 2.2 Add a `bin` entry:
    ```json
    "bin": {
      "lint": "dist/bin/lint.js"
    }
    ```

#### 🧰 Lint Command Implementation
- [x] 3. Create `src/bin/lint.ts`
- [x] 4. Follow structure and logic of `zeta.house/scripts/lint.mjs`
- [x] 5. Implement lint runners for:
  - [x] 5.1 `eslint` for JS/TS files
  - [x] 5.2 `prettier` for markdown
  - [x] 5.3 `shellcheck` for shell scripts
  - ~~[ ] 5.4 `clang-format` for C/C++~~
  - ~~[ ] 5.5 `nixfmt` for Nix files~~
  - ~~[ ] 5.6 `rustfmt` for Rust files~~
- [x] 6. Ensure the script:
  - [x] 6.1 Checks if each native dependency exists before executing
  - [x] 6.2 Uses a fast, cross-platform `commandExists` utility

#### 🧪 Add Utility Function
- [x] 7. Implement `commandExists(programName: string): boolean` that:
  - [x] 7.1 Works on Linux, macOS, and Windows
  - [x] 7.2 Is fast and lightweight
  - [x] 7.3 Returns `true`/`false` for command existence
- [x] 7A. We need this package to be extendable by downstream repos as well so they can extend the package and ignore certain rules or implement custom rules if need be.

#### 📦 Package Dependencies
- [x] 8. Add all relevant `eslint`, `@typescript-eslint`, `prettier`, and related dependencies
- [x] 9. Refer to `zeta.house` `package.json` as a baseline
- [x] 10. Do **not** include native tools like `shellcheck` or `clang-format` in `package.json`

#### 🔁 Plugin De-emphasis
- [x] 11. Do not export an ESLint plugin as the main export
- [x] 12. Optionally retain plugin code for internal use only

#### 🚀 Prepare for Release
- [ ] 13. Tag the repo with version `v0.0.1` (CI will handle publishing)

#### 🌱 Downstream Integration
- [ ] 14. Create a feature branch in `js-logger`
- [ ] 15. Integrate `@matrixai/lint` and verify it runs correctly
- [x] 16. Optionally filter directories or extend the lint command if needed

### Final checklist

* [ ] Domain specific tests
* [ ] Full tests
* [ ] Updated inline-comment documentation
* [ ] Lint fixed
* [ ] Squash and rebased
* [ ] Sanity check the final build

### CODE REVIEW

**Comment by CDeltakai** [comment]
Note: the current implmentation of this linter will ONLY lint files that are included inside configured tsconfigs. 

I've added a few extra features after testing this on Polykey-Enterprise. Repos with multiple tsconfigs (like Polykey-Enterprise) posed a bit of a problem since the linter would only pick up the closest tsconfig, or in this case the one in the root folder. At first I figured that I could search through the entire repository and look for any tsconfigs, however this proved too unsafe an approach since you can't know for sure what the user wants linted or not. Additionally, there was an annoying edge case with Polykey-Enterprise where the root tsconfig excluded the folders from src/app/ but the other tsconfig in that folder re-included those folders again, which would end up confusing the linter on what to lint and what to ignore.

I've decided that the solution to this is to make the user create a matrixai-lint-config.json which has 2 fields:

```
"tsconfigPaths": []
"forceInclude": []
```

Here, the user can explicitly state which tsconfigs they want to use (to recap, tsconfigs are required by the linter to perform type-aware linting e.g. no-floating-promises rule), and the forceInclude array will allow the user to override any excludes from their tsconfigs, though be aware that there still needs to be a tsconfig that includes those files otherwise there will be parsing error. Apparently projects with multiple tsconfigs and "conflicting" includes/excludes is not an uncommon thing so chances are this linter would run into more situations like this in the future necessitating this feature.

**Comment by CDeltakai** [comment]
The user can also override the internal config with their own by typing in the console ```matrixai-lint --user-config```, which will use the eslint.config.js in the root of the repo. They can then either eschew the internal config entirely or extend it by importing the config directly from @matrixai/lint/config like so:

```
//eslint.config.mjs
import matrixaiConfig from '@matrixai/lint/config';
export default [
...matrixaiConfig,
//the rest of your stuff
]
```
There is also the option to point to some config in any folder you want with ```matrixai-lint --config <path/to/config>``` if so desired.

File: README.md
Comment (CMCDragonkai) [inline README.md:1]: The README.md should be restructured to follow the other READMEs like js-logger.

File: src/configs/matrixai-config-bundle.ts [hunk 1]
```diff
@@ -0,0 +1,260 @@
...
+      'react/prop-types': 'off',
+
+      '@matrixai/no-aliased-imports': [
+        'error',
+        {
+          aliases: [{ prefix: '#', target: 'src' }],
+          includeFolders: ['src'],
+          autoFix: true,
+        },
+      ],
```
Comment (CMCDragonkai) [inline src/configs/matrixai-config-bundle.ts:105]: Let's get all of our "custom" rules at the VERY top of all rules.

File: src/configs/recommended.ts
Comment (CMCDragonkai) [inline src/configs/recommended.ts:1]: This is going to be removed?

File: src/configs/matrixai-config-bundle.ts
Comment (CMCDragonkai) [inline src/configs/matrixai-config-bundle.ts:1]: If we are going to have a `configs` directory. We can restructure this as individual "kinds" of config.

So right now we just have a `matrixai-typescript` codebase.

Right now that looks like:

```
application-js
application-js-cloudflare
library-js
native-library-js
```

Since we start with 1 config, we can just call it:

```
js.ts
```

File: src/runners/runESlint.ts
Comment (CMCDragonkai) [inline src/runners/runESlint.ts:1]: Just put this into your `src/utils.ts`.

File: src/utils/resolveMatrixConfig.ts
Comment (CMCDragonkai) [inline src/utils/resolveMatrixConfig.ts:1]: All of this could be under a single `src/utils.ts`.

File: eslint.config.mjs
Comment (CMCDragonkai) [inline eslint.config.mjs:1]: I would remove this file, and have the linter run itself on this same codebase.

File: package.json [hunk 2]
```diff
@@ -9,36 +9,54 @@
...
+  },
   "scripts": {
     "prepare": "tsc -p ./tsconfig.build.json",
-    "build": "shx rm -rf ./lib && tsc -p ./tsconfig.build.json",
+    "build": "shx rm -rf ./dist && tsc -p ./tsconfig.build.json && shx chmod +x dist/bin/lint.js",
     "postversion": "npm install --package-lock-only --ignore-scripts --silent",
     "tsx": "tsx",
-    "test": "node ./scripts/test.mjs",
     "lint": "eslint '{src,tests,scripts,benches}/**/*.{js,mjs,ts,mts,jsx,tsx}'",
     "lintfix": "eslint '{src,tests,scripts,benches}/**/*.{js,mjs,ts,mts,jsx,tsx}' --fix",
```
Comment (CMCDragonkai) [inline package.json:30]: These 2 should be updated to running the `lint` command.

Comment (CMCDragonkai) [reply package.json:30]: Actually `matrixai-lint` command.

Comment (CMCDragonkai) [reply package.json:30]: It would only work after you do a `npm run build`.

File: src/utils/resolveMatrixConfig.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
+// Utils/resolveMatrixConfig.ts
```
Comment (CMCDragonkai) [inline src/utils/resolveMatrixConfig.ts:1]: Get rid of these comments.

File: src/utils/resolveMatrixConfig.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+      const text = fs.readFileSync(cfgPath, 'utf8').trim();
+      rawCfg = text.length ? JSON.parse(text) : {};
+    } catch {
+      console.error(
+        '[matrixai‑lint]  ✖  matrixai-lint-config.json is not valid JSON – falling back to defaults.',
+      );
+    }
+  }
+
+  // ---------- helpers ----------
```
Comment (CMCDragonkai) [inline src/utils/resolveMatrixConfig.ts:31]: Get rid of these vestigial comments.

**Comment by CMCDragonkai** [comment]
Let's get this merged after fixing those minor things, and then integrate into js-logger as a start.

**Comment by CMCDragonkai** [comment]
Request @tegefaulkes to review once you are finished cleaning up this codebase in relation to conventions. Make sure to review js-logger as a template for how you should structure things, finish up the above reviews accordingly.

File: src/utils/utils.ts [hunk 5]
```diff
@@ -0,0 +1,165 @@
...
+
+  const toStringArray = (v: unknown): string[] =>
+    typeof v === 'string'
+      ? [v]
+      : Array.isArray(v)
+        ? v.filter((x): x is string => typeof x === 'string')
+        : [];
+
+  const cfg = rawCfg as { tsconfigPaths?: unknown; forceInclude?: unknown };
+
```
Comment (tegefaulkes) [inline src/utils/utils.ts:105]: Don't do this, You're throwing type checking out of the window. It's also kinda pointless? You're saying an arbitray JSON object May have these keys and they are unknown? That's already true.

File: src/utils/utils.ts [hunk 6]
```diff
@@ -0,0 +1,165 @@
...
+      );
+    }
+  }
+
+  const toStringArray = (v: unknown): string[] =>
+    typeof v === 'string'
+      ? [v]
+      : Array.isArray(v)
+        ? v.filter((x): x is string => typeof x === 'string')
+        : [];
```
Comment (tegefaulkes) [inline src/utils/utils.ts:102]: This is weird and confusing to read. Its better to make a separate utility function for it and write it out cleaner with proper documentation.

Besides that it seems to be unnecessary. I'm assuming you're creating a custom config structure? Or is it based on something? If you define it then why are you having a field be optionally a single string or an array of strings? Just but the lowest common denominator of an array of strings to avoid the weird logic.

File: src/utils/utils.ts (hunk 5 above)
Comment (tegefaulkes) [reply src/utils/utils.ts:105]: This is is a json file with an expected format then we need proper validation. Polykey already uses `https://www.npmjs.com/package/ajv` to apply schema to it's status file. We can use the same thing here.

This may be another add on issue.

File: src/utils/utils.ts [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+    if (exists(rootTs)) tsconfigPaths.push(rootTs);
+  }
+
+  return { tsconfigPaths, forceInclude };
+}
+
+interface Patterns {
+  files: string[];
+  ignore: string[];
+}
```
Comment (tegefaulkes) [inline src/utils/utils.ts:130]: Don't use interfaces for types, just define a new type. But you don't even need to do that when it's only used once. In that case just define the return type directly like this.

```ts
function buildPatterns(
  tsconfigPath: string,
  forceInclude: string[] = [],
): {
  files: string[];
  ignore: string[];
}
```

If you had to use a type multiple times then define it with 

```ts
type Pattern = {
  files: string[];
  ignore: string[];
};
```

And if this is used across files the define it in a `types.ts` file.

File: tests/index.test.ts
Comment (tegefaulkes) [inline tests/index.test.ts:1]: Why did we add tests files if there is nothing to test? Was it to appease the CI? This repo will really have nothing to test so I don't see a need for any jest code or it's dependencies in `package.json`.

Comment (CDeltakai) [reply tests/index.test.ts:1]: Yes, from what I remember if we didn't add the tests, even if they were just empty, the CI would constantly fail.

File: src/utils/utils.ts [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+  const abs = (p: string) => path.resolve(repoRoot, p);
+  const exists = (p: string) => fs.existsSync(p);
+
+  let rawCfg: unknown = {};
+
+  if (exists(cfgPath)) {
+    try {
+      const text = fs.readFileSync(cfgPath, 'utf8').trim();
+      rawCfg = text.length ? JSON.parse(text) : {};
+    } catch {
```
[the remark this answers is not in the document]
Comment (CDeltakai) [reply src/utils/utils.ts:81]: > Besides that why are you checking it. If you can't parse the JSON isn't that explicitly an error?
> 

So the main reason is that I wanted the linter to just ignore the invalid JSON and just fall back to defaults if possible. Now that I think about it, if a user has decided to have a matrixai-lint-config.json to begin with, then they probably don't want to use the defaults anyway, so it might be better to just hard fail in that case so the user is forced to fix it or not have it at all.

File: src/bin/lint.ts [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+        '--user-config given but no local ESLint config was found. Falling back to built-in config.',
+      );
+    }
+  }
+
+  console.error('Running eslint:');
+  runESLint({ fix, configPath: chosenConfig }).catch((err) => {
+    console.error('ESLint failed:', err);
+    process.exit(1);
+  });
```
[the remark this answers is not in the document]
Comment (CDeltakai) [reply src/bin/lint.ts:65]: I'm not entirely sure why, but if I await runESLint, it seems to prevent the shellcheck linting and Prettier from running, which is why I switched to this instead.

[the remark this answers is not in the document]
Comment (CDeltakai) [reply src/bin/lint.ts]: Nevermind, I realized that it was because the runESLint function was exiting the process prematurely if any error was found, causing the other parts to not run unless there was no linting errors at all.

File: src/utils/utils.ts (hunk 6 above)
Comment (CDeltakai) [reply src/utils/utils.ts]: Well I suppose I figured it would make things slightly easier on the user if they could just write a single string without needing to always have the square brackets if they really only had one thing they wanted to put in either field. 

I did extract the toStringArray as its own function with simpler syntax so it doesn't look as opaque though.

File: src/runners/runESlint.ts
[the remark this answers is not in the document]
Comment (CDeltakai) [reply src/runners/runESlint.ts:1]: So originally I thought that the runESLint.ts would have to be some kind of background runner to get ESLint to work externally, but after awhile it became clear that it didn't really need to be that way and it worked fine as just a standalone function.

I've extracted it and just placed it into the utils.ts file and removed the folder.

File: src/utils/utils.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (CDeltakai) [reply src/utils/utils.ts]: > Also it's probably better to use a dynamic import `await import(cfgPath, { assert: { type: 'json' } })`

Is this strictly necessary? If I do change to use this that would turn the function asynchronous which would require a number of refactors in the rest of the code. It's not a huge amount of work of course so if you insist I can make the change, it just seems to me that the current manner works fine and won't require any refactors.

File: src/utils/utils.ts (hunk 5 above)
Comment (CDeltakai) [reply src/utils/utils.ts]: > Don't do this, You're throwing type checking out of the window. It's also kinda pointless? You're saying an arbitray JSON object May have these keys and they are unknown? That's already true.

I've changed this to use a partial type with RawMatrixCfg:

```
type RawMatrixCfg = Partial<{
  tsconfigPaths: unknown;
  forceInclude: unknown;
}>; 
```
and then casting it to this type with the ```asRawMatrixCfg``` method as a basic type guard. I think this will fix this issue in the short term but I'll keep this open in case we want a proper issue with the proper validation fix.
